libpromises/mod_custom.c

```diff
@@ -0,0 +1,741 @@
+    if (global < LOG_LEVEL_ERR)
+    {
+        assert((global == LOG_LEVEL_NOTHING) || (global == LOG_LEVEL_CRIT));
+        return LogLevelToString(LOG_LEVEL_ERR);
+    }
+    return LogLevelToString(global);
+}
+
+static PromiseResult PromiseModule_Evaluate(
+    PromiseModule *module, EvalContext *ctx, const Promise *pp)
+{
+    assert(module != NULL);
+    assert(pp != NULL);
+
+    PromiseModule_AppendString(module, "operation", "evaluate_promise");
+    PromiseModule_AppendString(
+        module, "log_level", LogLevelToRequestFromModule());
+    PromiseModule_AppendString(module, "promiser", pp->promiser);
+
+    PromiseModule_AppendAllAttributes(module, pp);
+    PromiseModule_Send(module);
+
+    JsonElement *response = PromiseModule_Receive(module);
+    PromiseResult result;
+    const char *const result_str = JsonObjectGetAsString(response, "result");
+
+    Attributes a = GetClassContextAttributes(ctx, pp); // TODO: WTF
+
+    if (result_str == NULL)
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "Promise module did not return a result for promise evaluation (bug in promise module: %s)",
+            module->path);
+    }
+    else if (StringEqual(result_str, "kept"))
+    {
+        result = PROMISE_RESULT_NOOP;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "not_kept"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was not kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "repaired"))
+    {
+        result = PROMISE_RESULT_CHANGE;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was repaired by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "error"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "An unexpected error occured in promise module (bug in promise module: %s)",
+            module->path);
+    }
+    else
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "Promise module returned unacceptable result: '%s' (bug in promise module: %s)",
```

Comment:
Same here and above regarding the promiser.

Reply:
Thanks, fixed.